### PR TITLE
[WIP] Use melpa version of magit-gh-pulls

### PR DESCRIPTION
**Pull request by robbyoconnor**
magit-gh-pulls is now on [melpa](http://melpa.org/#/magit-gh-pulls) -- There are a few kinks that need to be worked out (hence the WIP).

### CODE REVIEW

**Comment by robbyoconnor** [comment]
I am getting this when using the melpa version: 
```
byte-code: Key sequence j q starts with non-prefix key j

```

**Comment by bmag** [comment]
@robbyoconnor [this](https://github.com/sigma/magit-gh-pulls/blob/master/magit-gh-pulls.el#L420) is probably the offending line in `magit-gh-pulls`:
```elisp
(define-key magit-status-mode-map (kbd "jq") 'magit-jump-to-pulls)
```
IIUC, the trouble stems from the evilification of `magit-status-mode-map`. Normally in magit-status, j-prefixed keys jump to different section, i.e. <kbd>js</kbd> (jump to staged section) and <kbd>ju</kbd> (jump to unstaged section). The evilification moves <kbd>j</kbd> to <kbd>J</kbd>, so jump keys are <kbd>Js</kbd>, <kbd>Ju</kbd>, etc. <kbd>j</kbd> is bound to `evil-visual-next-line`, which really isn't a prefix key, so the above `define-key` snippet fails.

If the `"jq"` in `magit-gh-pulls` could be customized (with `nil` meaning "don't bind `magit-jump-to-pulls`"), we could avoid the error. Worth opening an issue/PR upstream.

Also, I expected the following snippet would work, but for some reason it doesn't. Probably an issue with evilification.
```elisp
(define-key magit-status-mode-map (kbd "Jq") 'magit-jump-to-pulls)
```

File: contrib/!source-control/github/packages.el [hunk 1]
```diff
@@ -86,3 +87,25 @@
         "gfC" 'spacemacs/git-link-commit-copy-url-only)
       ;; default is to open the generated link
       (setq git-link-open-in-browser t))))
+
+(defun github/init-magit-gh-pulls ()
+  (use-package magit-gh-pulls
+    :commands magit-gh-pulls-mode
```
Comment (bmag) [inline contrib/!source-control/github/packages.el]: Isn't this line redundant? When using the melpa version, `magit-gh-pulls-mode` is already autoloaded.

Comment (robbyoconnor) [inline contrib/!source-control/github/packages.el]: I literally copied from @syl20bnr's extension

**Comment by robbyoconnor** [comment]
Shit did I just merge this :(

File: contrib/!source-control/github/packages.el (hunk 1 above)
Comment (syl20bnr) [inline contrib/!source-control/github/packages.el]: The MELPA package comes with autoloads, which is not the case when this is an extension. So you can change it to `:defer t`.

**Comment by robbyoconnor** [comment]
Literally nothing @bmag has proposed works -- maybe it's just i'm not in the mindset to work on this -- things sort of work now -- so let's leave this sit for awhile -- maybe they'll fix upstream soon.

File: contrib/!source-control/github/packages.el (hunk 1 above)
Comment (robbyoconnor) [inline contrib/!source-control/github/packages.el]: This issue aside, -- there's still a problem with the upstream code not playing nice :|

**Comment by syl20bnr** [comment]
We cannot ask upstream to make a modification just for spacemacs, the default `jq` bindings makes sense and so we have to find another way to make the package work.

I think that if we `eval-after-load` the binding on `j` after the loading of `magit-gh-pulls` (when it is used) we should be able to prevent this issue.

**Comment by robbyoconnor** [comment]
Let me know what needs to be changed; I do think this should be merged; just need to figure out how to make it work :)

**Comment by robbyoconnor** [comment]
I'm out of ideas myself

**Comment by cmccloud** [comment]
@robbyoconnor I don't know if this will work, but it seems to me that what we want to do is initialize magit-gh-pulls before evilification of magit maps occurs. I *think* that would mean that evilification would take care of the rebinding for us. Maybe something like this?
```
(setq github-packages
      '(
        gist
        git-link
        github-browse-file
        magit-gh-pulls
        magit
        ;; not up to date
        ;; helm-gist
        ))

;; omitted code

(defun github/pre-init-magit ()
  (spacemacs|use-package-add-hook magit
    :pre-config
    (progn
      (use-package magit-gh-pulls
        :init
        (progn
          (defun spacemacs/load-gh-pulls-mode ()
            "Start `magit-gh-pulls-mode' only after a manual request."
            (interactive)
            (magit-gh-pulls-mode)
            (magit-gh-pulls-reload))
          (defun spacemacs/fetch-gh-pulls-mode ()
            "Start `magit-gh-pulls-mode' only after a manual request."
            (interactive)
            (magit-gh-pulls-mode)
            (magit-gh-pulls-fetch-commits)))
        :config
        (spacemacs|diminish magit-gh-pulls-mode "Github-PR")))))

(defun github/init-magit-gh-pulls ())
```
My understanding is that the :pre-config block above will only be triggered when the :config block in init-magit is ready to run, i.e. after magit has been loaded.

Assuming it works, maybe @syl20bnr could give us his opinion.

**Comment by cmccloud** [comment]
I just made a custom layer with the above code and it seemed to work - Jq from a magit-status buffer worked correctly and there was no byte-code warning

**Comment by TheBB** [comment]
That's some serious trickery going on, there.

**Comment by cmccloud** [comment]
There's probably a cleaner way to do it, but I couldn't think of anything else less tricky.
I am quite pleased with the resolution of the spacemacs package management system - in this case it really wasn't that tough to put code exactly where it needed to go.

**Comment by robbyoconnor** [comment]
@cmccloud -- I gave up personally -- if you make a PR -- I'll close this

**Pull request closed without merging**